### PR TITLE
feat: add ECSContainerInstanceEvent

### PR DESCRIPTION
**Pull request by wafuwafu13**
*Issue #, if available:*

*Description of changes:*
ref: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_cwe_events.html#ecs_container_instance_events

I verified in my environment.

<details>
<summary>Lambda code</summary>

```go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/aws/aws-lambda-go/lambda"
)

type ECSContainerInstanceEvent struct {
	Version    string                              `json:"version"`
	ID         string                              `json:"id"`
	DetailType string                              `json:"detail-type"`
	Source     string                              `json:"source"`
	Account    string                              `json:"account"`
	Time       time.Time                           `json:"time"`
	Region     string                              `json:"region"`
	Resources  []string                            `json:"resources"`
	Detail     ECSContainerInstanceEventDetailType `json:"detail"`
}

type ECSContainerInstanceEventDetailType struct {
	AgentConnected       bool        `json:"agentConnected"`
	Attributes           []Attribute `json:"attributes"`
	ClusterArn           string      `json:"clusterArn"`
	ContainerInstanceArn string      `json:"containerInstanceArn"`
	EC2InstanceID        string      `json:"ec2InstanceId"`
	RegisteredResources  []Resource  `json:"registeredResources"`
	RemainingResources   []Resource  `json:"remainingResources"`
	Status               string      `json:"status"`
	Version              int         `json:"version"`
	VersionInfo          VersionInfo `json:"versionInfo"`
	UpdatedAt            time.Time   `json:"updatedAt"`
}

type Attribute struct {
	Name string `json:"name"`
}

type Resource struct {
	Name           string    `json:"name"`
	Type           string    `json:"type"`
	IntegerValue   int       `json:"integerValue,omitempty"`
	StringSetValue []*string `json:"stringSetValue,omitempty"`
}

type VersionInfo struct {
	AgentHash     string `json:"agentHash"`
	AgentVersion  string `json:"agentVersion"`
	DockerVersion string `json:"dockerVersion"`
}

func (r Resource) MarshalJSON() ([]byte, error) {
	type Alias Resource
	aux := struct {
		StringSetValue json.RawMessage `json:"stringSetValue,omitempty"`
		Alias
	}{
		Alias: (Alias)(r),
	}

	if r.StringSetValue != nil {
		b, err := json.Marshal(r.StringSetValue)
		if err != nil {
			return nil, err
		}
		aux.StringSetValue = b
	}

	return json.Marshal(&aux)
}

func handler(ctx context.Context, ecsEvent ECSContainerInstanceEvent) {
	outputJSON, _ := json.MarshalIndent(ecsEvent, "", " ")
	fmt.Printf("output: %s", outputJSON)
}

func main() {
	lambda.Start(handler)
}
```
</details>

Change status to `DRAINING`

<details>
<summary>Cloudwatch Logs</summary>

```txt
  | 2023-04-16T12:59:37.095+01:00 | START RequestId: <START RequestId> Version: $LATEST
  | 2023-04-16T12:59:37.096+01:00 | err: <nil>
  | 2023-04-16T12:59:37.096+01:00 | output: {
  | 2023-04-16T12:59:37.096+01:00 | "version": "0",
  | 2023-04-16T12:59:37.096+01:00 | "id": "<id>",
  | 2023-04-16T12:59:37.096+01:00 | "detail-type": "ECS Container Instance State Change",
  | 2023-04-16T12:59:37.096+01:00 | "source": "aws.ecs",
  | 2023-04-16T12:59:37.096+01:00 | "account": "<account>",
  | 2023-04-16T12:59:37.096+01:00 | "time": "2023-04-16T11:59:36Z",
  | 2023-04-16T12:59:37.096+01:00 | "region": "eu-west-1",
  | 2023-04-16T12:59:37.096+01:00 | "resources": [
  | 2023-04-16T12:59:37.096+01:00 | "<resources>"
  | 2023-04-16T12:59:37.096+01:00 | ],
  | 2023-04-16T12:59:37.096+01:00 | "detail": {
  | 2023-04-16T12:59:37.096+01:00 | "agentConnected": false,
  | 2023-04-16T12:59:37.096+01:00 | "attributes": [
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "name": "ecs.capability.secrets.asm.environment-variables"
  | 2023-04-16T12:59:37.096+01:00 | },
...
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "name": "com.amazonaws.ecs.capability.task-iam-role"
  | 2023-04-16T12:59:37.096+01:00 | }
  | 2023-04-16T12:59:37.096+01:00 | ],
  | 2023-04-16T12:59:37.096+01:00 | "clusterArn": "<clusterArn>",
  | 2023-04-16T12:59:37.096+01:00 | "containerInstanceArn": "<containerInstanceArn>",
  | 2023-04-16T12:59:37.096+01:00 | "ec2InstanceId": "<ec2InstanceId>",
  | 2023-04-16T12:59:37.096+01:00 | "registeredResources": [
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "name": "CPU",
  | 2023-04-16T12:59:37.096+01:00 | "type": "INTEGER",
  | 2023-04-16T12:59:37.096+01:00 | "integerValue": 1024
  | 2023-04-16T12:59:37.096+01:00 | },
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "name": "MEMORY",
  | 2023-04-16T12:59:37.096+01:00 | "type": "INTEGER",
  | 2023-04-16T12:59:37.096+01:00 | "integerValue": 970
  | 2023-04-16T12:59:37.096+01:00 | },
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "stringSetValue": [
  | 2023-04-16T12:59:37.096+01:00 | "22",
  | 2023-04-16T12:59:37.096+01:00 | "2376",
  | 2023-04-16T12:59:37.096+01:00 | "2375",
  | 2023-04-16T12:59:37.096+01:00 | "51678",
  | 2023-04-16T12:59:37.096+01:00 | "51679"
  | 2023-04-16T12:59:37.096+01:00 | ],
  | 2023-04-16T12:59:37.096+01:00 | "name": "PORTS",
  | 2023-04-16T12:59:37.096+01:00 | "type": "STRINGSET"
  | 2023-04-16T12:59:37.096+01:00 | },
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "stringSetValue": [],
  | 2023-04-16T12:59:37.096+01:00 | "name": "PORTS_UDP",
  | 2023-04-16T12:59:37.096+01:00 | "type": "STRINGSET"
  | 2023-04-16T12:59:37.096+01:00 | }
  | 2023-04-16T12:59:37.096+01:00 | ],
  | 2023-04-16T12:59:37.096+01:00 | "remainingResources": [
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "name": "CPU",
  | 2023-04-16T12:59:37.096+01:00 | "type": "INTEGER",
  | 2023-04-16T12:59:37.096+01:00 | "integerValue": 1024
  | 2023-04-16T12:59:37.096+01:00 | },
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "name": "MEMORY",
  | 2023-04-16T12:59:37.096+01:00 | "type": "INTEGER",
  | 2023-04-16T12:59:37.096+01:00 | "integerValue": 970
  | 2023-04-16T12:59:37.096+01:00 | },
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "stringSetValue": [
  | 2023-04-16T12:59:37.096+01:00 | "22",
  | 2023-04-16T12:59:37.096+01:00 | "2376",
  | 2023-04-16T12:59:37.096+01:00 | "2375",
  | 2023-04-16T12:59:37.096+01:00 | "51678",
  | 2023-04-16T12:59:37.096+01:00 | "51679"
  | 2023-04-16T12:59:37.096+01:00 | ],
  | 2023-04-16T12:59:37.096+01:00 | "name": "PORTS",
  | 2023-04-16T12:59:37.096+01:00 | "type": "STRINGSET"
  | 2023-04-16T12:59:37.096+01:00 | },
  | 2023-04-16T12:59:37.096+01:00 | {
  | 2023-04-16T12:59:37.096+01:00 | "stringSetValue": [],
  | 2023-04-16T12:59:37.096+01:00 | "name": "PORTS_UDP",
  | 2023-04-16T12:59:37.096+01:00 | "type": "STRINGSET"
  | 2023-04-16T12:59:37.096+01:00 | }
  | 2023-04-16T12:59:37.096+01:00 | ],
  | 2023-04-16T12:59:37.096+01:00 | "status": "DRAINING",
  | 2023-04-16T12:59:37.096+01:00 | "version": 36,
  | 2023-04-16T12:59:37.096+01:00 | "versionInfo": {
  | 2023-04-16T12:59:37.096+01:00 | "agentHash": "<agentHash>",
  | 2023-04-16T12:59:37.096+01:00 | "agentVersion": "1.70.0",
  | 2023-04-16T12:59:37.096+01:00 | "dockerVersion": "DockerVersion: 20.10.17"
  | 2023-04-16T12:59:37.096+01:00 | },
  | 2023-04-16T12:59:37.096+01:00 | "updatedAt": "2023-04-16T11:59:36.836Z"
  | 2023-04-16T12:59:37.096+01:00 | }
  | 2023-04-16T12:59:37.097+01:00 | END RequestId: <END RequestId>
  | 2023-04-16T12:59:37.097+01:00 | REPORT RequestId: <REPORT RequestId> Duration: 1.65 ms Billed Duration: 2 ms Memory Size: 128 MB Max Memory Used: 31 MB
```

</details>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: events/ecs_container_instance.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	Type           string    `json:"type"`
+	IntegerValue   int       `json:"integerValue,omitempty"`
+	StringSetValue []*string `json:"stringSetValue,omitempty"`
+}
+
+type VersionInfo struct {
+	AgentHash     string `json:"agentHash"`
+	AgentVersion  string `json:"agentVersion"`
+	DockerVersion string `json:"dockerVersion"`
+}
```
Comment (bmoffatt) [inline events/ecs_container_instance.go:49]: `events` is too global a namespace for generic types like these. They should be prefixed like `ECSContainerInstanceEvent<Foo>`

File: events/ecs_container_instance.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+type Attribute struct {
+	Name string `json:"name"`
+}
+
+type Resource struct {
+	Name           string    `json:"name"`
+	Type           string    `json:"type"`
+	IntegerValue   int       `json:"integerValue,omitempty"`
+	StringSetValue []*string `json:"stringSetValue,omitempty"`
```
Comment (bmoffatt) [inline events/ecs_container_instance.go:42]: should it be `[]string`? `[]*string` implies the array is expected to contain nulls

File: events/ecs_container_instance_test.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (bmoffatt) [inline events/ecs_container_instance_test.go:1]: 2023

File: events/ecs_container_instance.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+type VersionInfo struct {
+	AgentHash     string `json:"agentHash"`
+	AgentVersion  string `json:"agentVersion"`
+	DockerVersion string `json:"dockerVersion"`
+}
+
+// MarshalJSON implements cuustom marshaling to marshal the struct into JSON format while preserving an empty string slice in `StringSetValue` field.
+func (r Resource) MarshalJSON() ([]byte, error) {
```
Comment (bmoffatt) [inline events/ecs_container_instance.go:52]: At first glance I thought this was overkill for the sake of the serde test

Needs follow-up to lookup schema in EventBridge Schema Registry. If the schema says this is never `null`, then yeah this is how it should be done!

File: events/ecs_container_instance.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	StringSetValue []*string `json:"stringSetValue,omitempty"`
+}
+
+type VersionInfo struct {
+	AgentHash     string `json:"agentHash"`
+	AgentVersion  string `json:"agentVersion"`
+	DockerVersion string `json:"dockerVersion"`
+}
+
+// MarshalJSON implements cuustom marshaling to marshal the struct into JSON format while preserving an empty string slice in `StringSetValue` field.
```
Comment (bmoffatt) [inline events/ecs_container_instance.go:51]: typo `cuustom` -> `custom`

File: events/ecs_container_instance.go (hunk 4 above)
Comment (wafuwafu13) [reply events/ecs_container_instance.go]: The aws.ecs@ECSContainerInstanceStateChange schema says nullable, so I remove this function. https://github.com/aws/aws-lambda-go/pull/502/commits/f9ac9df1c93423728f2c29df03f9dd6d4ce73298
ref:
https://eu-west-1.console.aws.amazon.com/events/home?region=eu-west-1#/registries/aws.events/schemas/aws.ecs%40ECSContainerInstanceStateChange

>"stringSetValue": {
            "type": "array",
            "items": {
              "type": "string",
              "nullable": true
            }
          },

File: events/ecs_container_instance.go [hunk 6]
```diff
@@ -36,10 +36,10 @@ type ECSContainerInstanceEventAttribute struct {
 }
 
 type ECSContainerInstanceEventResource struct {
-	Name           string    `json:"name"`
-	Type           string    `json:"type"`
-	IntegerValue   int       `json:"integerValue,omitempty"`
-	StringSetValue []*string `json:"stringSetValue,omitempty"`
```
Comment (bmoffatt) [inline events/ecs_container_instance.go]: I think the schema says you got this right the first time around :) https://github.com/aws/aws-lambda-go/pull/502#discussion_r1174432104

File: events/ecs_container_instance.go (hunk 2 above)
Comment (bmoffatt) [reply events/ecs_container_instance.go]: see other comment: https://github.com/aws/aws-lambda-go/pull/502#discussion_r1174654102 I think schema say `[]*string` was the correct way

File: events/ecs_container_instance.go (hunk 6 above)
Comment (wafuwafu13) [reply events/ecs_container_instance.go]: :+1: git reset https://github.com/aws/aws-lambda-go/commit/7316ef154a6ae107e075ea453f5133615290b665

**Review by bmoffatt — APPROVED** [review]

**Pull request merged**